kong/router/atc.lua

```diff
@@ -382,12 +383,19 @@ function _M:matching(params)
 
   local request_prefix = matched_route.strip_path and matched_path or nil
 
+  local uri_captures = nil
+  -- The whole matched uri will be put in captures[0], so we need to check
+  -- the number of table keys >= 2 to determine if there is any uri captured.
+  if captures and tb_nkeys(captures) >= 2 then
```

Comment:
The function `tb_nkeys` is an `O(n)` operation, and it is in the hot proxy path. I‘m concerned about the performance impact.

The [atc-router knows the length of this table](https://github.com/Kong/atc-router/blob/ffd11db657115769bf94f0c4f915f98300bc26b6/lib/resty/router/context.lua#L116), getting it from that could get performance some benefits.

Reply:
@chronolaw @StarlightIbuki What do you think?

Reply:
@ADD-SP Sure, it might be fine since the captures should be generally a small table.

Reply:
@ADD-SP I have a simpler solution: set the `captures[0]` to zero and then check with `next(captures)`.

Reply:
> @ADD-SP I have a simpler solution: set the captures[0] to nil and then check with next(captures).

This is hard to understand.

Reply:
> @ADD-SP I have a simpler solution: set the `captures[0]` to nil and then check with `next(captures)`.

It is a little hacky.

Reply:
Performance is more important than straightforwardness, especially on hot proxy path. For future maintainers, comments are what provides context.

Reply:
I saw @chronolaw has submitted a PR to return the capture count. Will change the code after it's merged.

Reply:
Will this work, I think the intention is reasonably clear?

`uri_captures = (captures and captures[1] or next(captures, 1)) and captures or nil`

Reply:
I think there are 2 options here:
1. Do nothing and go with what we have here
2. Extend ATC Router

Regarding option 1:
Even if `tb_nkeys` is an `O(n)` operation it does not increase the overall time complexity of this function as it's generally `O(n)`. The next question is how many other `O(n)` operations come here into play so that we could estimate the impact of adding this one. From what I see there are at least 2 other `O(n)` calls: 
- here: https://github.com/Kong/atc-router/blob/main/lib/resty/router/context.lua#L118
- and here: https://github.com/Kong/atc-router/blob/main/src/ffi.rs#L246

So adding a third one will decrease the performance here by ~33% with regard to captures length. Now I reckon - realistically the captures length _shouldn't_ be that long. The URLs are usually at most around 2k characters - maybe up to 8k so in the worst case scenario we have 8k captures.
Now having said all that - I think it might not be an issue but on the other hand it's a hot path and every bit matters so I think we'd need to perf test it.

Regarding option 2:
The captures length is _right there_. If it's helpful we can just return with captures table: https://github.com/Kong/atc-router/blob/main/lib/resty/router/context.lua#L115

Reply:
We can probably just remove this check, as `captures` will naturally be `nil` if no regex match has taken place.

Reply:
> I agree with dndx. we could remove this logic ( if captures ... end) and let  uri_captures = captures.

I cannot get this point. It returns captures[0] = {whole match patch} even if no regex matched, so it won't be nil without changing the ATC router.

Reply:
@dndx @chronolaw @StarlightIbuki @ADD-SP @nowNick I have updated this patch with `next` and it should remove the O(N) operation for now, would you mind taking a look again?

Reply:
@chronolaw @dndx @ADD-SP @StarlightIbuki @nowNick  Is it good to change to 


```lua
-- The whole match path will be placed in captures if it's matched,
-- and we need to check if any other fields inside captures.
-- The `captures[1] or captures[0] and next(captures, 0)` is to check if any other fields except captures[0]. 
uri_captures = (captures and (captures[1] or captures[0] and next(captures, 0))) and captures or nil
```

Reply:
We are still in discussion, just give us more time, thanks.

Reply:
> @ADD-SP I have a simpler solution: set the captures[0] to nil and then check with next(captures).

@StarlightIbuki I think we can call `next` twice, it should be enough, and don't forget to add a comment to explain it in detail.